### PR TITLE
fix(typescript-estree): support `declare` and `export` modifiers for nested namespace declaration

### LINKED ISSUE #7409: Bug: `typescript-estree` does not support `declare namespace foo.bar {}`
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEA7KBbEBnADlMCBmA9gQHQBGs8A3vAL5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
declare namespace foo.bar {}
```

### ESLint Config

_No response_

### tsconfig

_No response_

### Expected Result

```diff
{
  "type": "TSModuleDeclaration",
  "kind": "namespace",
  "body": {
    "type": "TSModuleBlock",
    "body": []
  },
- "declare": false,
+ "declare": true,
  "global": false,
  "id": {
    "left": {
      "decorators": [],
      "name": "astroHTML",
      "optional": false,
      "type": "Identifier"
    },
    "right": {
      "decorators": [],
      "name": "JSX",
      "optional": false,
      "type": "Identifier"
    },
    "type": "TSQualifiedName"
  }
}
```

### Actual Result

```diff
{
  "type": "TSModuleDeclaration",
  "kind": "namespace",
  "body": {
    "type": "TSModuleBlock",
    "body": []
  },
  "declare": false,
  "global": false,
  "id": {
    "left": {
      "decorators": [],
      "name": "astroHTML",
      "optional": false,
      "type": "Identifier"
    },
    "right": {
      "decorators": [],
      "name": "JSX",
      "optional": false,
      "type": "Identifier"
    },
    "type": "TSQualifiedName"
  }
}
```

### Additional Info

context: https://github.com/prettier/prettier/issues/15207

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `X.Y.Z` |
| `@typescript-eslint/parser`            | `X.Y.Z` |
| `@typescript-eslint/rule-tester`       | `X.Y.Z` |
| `@typescript-eslint/scope-manager`     | `X.Y.Z` |
| `@typescript-eslint/typescript-estree` | `6.2.1` |
| `@typescript-eslint/type-utils`        | `X.Y.Z` |
| `@typescript-eslint/utils`             | `X.Y.Z` |
| `TypeScript`                           | `X.Y.Z` |
| `ESLint`                               | `X.Y.Z` |
| `node`                                 | `X.Y.Z`

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7409 
- [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Support following syntax:

```ts
declare namespace foo.bar {}
export namespace foo.bar {}
export declare namespace foo.bar {}
```

### CODE REVIEW

File: packages/ast-spec/src/declaration/TSModuleDeclaration/fixtures/namespace-nested-once-declare/snapshots/1-TSESTree-AST.shot [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        type: "TSModuleBlock",
+        body: [],
+
+        range: [26, 28],
+        loc: {
+          start: { column: 26, line: 1 },
+          end: { column: 28, line: 1 },
+        },
+      },
+      declare: false,
```
Comment (bradzacher) [inline packages/ast-spec/src/declaration/TSModuleDeclaration/fixtures/namespace-nested-once-declare/snapshots/1-TSESTree-AST.shot:19]: your fix has not worked.

**Review by bradzacher — CHANGES_REQUESTED** [review]
thanks for looking into this!
looks like the logic isn't right still

File: packages/ast-spec/src/declaration/TSModuleDeclaration/fixtures/namespace-nested-once-declare/snapshots/1-TSESTree-AST.shot (hunk 1 above)
Comment (sosukesuzuki) [reply packages/ast-spec/src/declaration/TSModuleDeclaration/fixtures/namespace-nested-once-declare/snapshots/1-TSESTree-AST.shot]: Thank you. I've fixed [f43949d](https://github.com/typescript-eslint/typescript-eslint/pull/7410/commits/f43949de5a55bb6ab96878d496dac5eeb88c9638) and updated snapshots [e8463d2](https://github.com/typescript-eslint/typescript-eslint/pull/7410/commits/e8463d276b42d8621c1f56647b3abf6006f0c806)

**Comment by EchoZhaoH** [comment]
The `export` keyword also needs to support

**Comment by bradzacher** [comment]
Oof yeah it looks like our handling is pretty shoddy overall:

```ts
declare namespace foo1 { } // ✅ declare: true
declare module foo2 { } // ✅ declare: true

declare namespace foo3.bar { } // ❌ declare: false
declare module foo4.def { } // ✅ declare: true

export declare namespace foo5 { } // ✅ export: "true" ✅ declare: true
export declare module foo6 { } // ✅ export: "true" ✅ declare: true

export declare namespace foo7.bar { } // ❌ export: "false" ❌ declare: false
export declare module foo8.def { } // ✅ export: "true" ✅ declare: true
```

If possible - can we add cases for and fix all of this at once?

**Comment by sosukesuzuki** [comment]
Hmm, codecov is also failing in the part that should have been tested. Am I wrong..?

**Comment by bradzacher** [comment]
nah our codecov setup just isn't able to collect coverage across packages right now, sadly.

**Review by bradzacher — APPROVED** [review]
thanks so much for fixing this - this fix LGTM!

**Pull request merged**